README.md
# Linear Actuator Controller

これは"いしかわきょーすけ"さん作のM5Stack ATOM用直動機構を動かすための非公式ライブラリです。


---

## Project Status

<div style="display: flex">
    <div style="width:1em; background-color: red;margin-right:1em;"></div>
    <span style="">This project will not be refurbished unless requested.</span>
</div>


## Description


### parts list

| No. |          | Device type. | Device                    | URL                                          |
| --- | -------- | ------------ | ------------------------- | -------------------------------------------- |
| 1.  | Required | Board        | ATOM Lite                 |                                              |
| 2.  | Required | Module       | M5Stack ATOM用1軸直動機構 | https://www.switch-science.com/products/7675 |


## Installation

This system is compiled using PlatformIO. Please install the extension from VS Code's extension recommendations. If you want to know more, check [here](https://docs.platformio.org/en/latest/).

* platformio.ini


```ini
[env:LinearActuatorController]
platform = espressif32
board = m5stack-atom
framework = arduino
monitor_speed = 115200
lib_deps =
    m5stack/M5Atom@^0.1.0
    fastled/FastLED@^3.5.0
    https://github.com/MasterAkari/LinearActuatorController.git
```

* main.cpp

```c++
#include "linear_actuator_controller.hpp"

#include <Arduino.h>
#include <Wire.h>

LinearActuatorController lac;

void setup()
{
    lac.begin();
}

void loop()
{
  // move position : 0[%]
  lac.move(0);
  delay(1000);
  // move position : 100[%]
  lac.move(100);
  delay(1000);
}
```

## Usage

* SPIFFSに情報を格納する予定です。情報をフォーマットすることがあります。

## Requirement

This system uses the following libraries.

* [m5stack/M5Atom](https://github.com/m5stack/M5Atom?utm_source=platformio&utm_medium=piohome)
* [fastled/FastLED](https://github.com/Makuna/NeoPixelBus?utm_source=platformio&utm_medium=piohome)

## Changelog

It is listed [here](./Changelog).

## Notes

* This system is not guaranteed to work.

## Support

* Not Supported.

## Roadmap

It is listed [here](./Changelog).

<!--
* Update irregularly.

<div style="display: flex">
    <div style="width:1em; background-color: red;margin-right:1em;"></div>
    <span style="">Next Update&nbsp;:&nbsp; </span>
    <span style="">No plans.</span>
</div>

-->

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.
Please make sure to update tests as appropriate.

## Authors and acknowledgment

Show your appreciation to those who have contributed to the project.

## License

[MIT License](./LICENSE)

---

<style>
  body {
    counter-reset: chapter;
}

h2 {
    counter-reset: sub-chapter;
}

h3 {
    counter-reset: section;
}

h4 {
    counter-reset: indexlist;
}

h1::before {
    counter-reset: chapter;
}

h2::before {
    counter-increment: chapter;
    content: counter(chapter) ". ";
}

h3::before {
    counter-increment: sub-chapter;
    content: counter(chapter) "-" counter(sub-chapter) ". ";
}

h4::before {
    counter-increment: section;
    content: counter(chapter) "-" counter(sub-chapter) "-" counter(section) ". ";
}

h5::before {
    counter-increment: indexlist;
    content: "(" counter(indexlist) ") ";
}

#sidebar-toc-btn {
    bottom: unset;
    top: 8px;
}

.markdown-preview.markdown-preview {
    h2 {
        border-bottom: 4px solid #eaecef;
    }

    h3 {
        border-bottom: 1px solid #eaecef;
    }
}

.md-sidebar-toc.md-sidebar-toc {
    padding-top: 40px;
}

#sidebar-toc-btn {
    bottom: unset;
    top: 8px;
}
</style>


Changelog
# Changes Overview

---

## Ver.0.2.0

* ペンプロッタ拡張キット用のクラスの追加

## Ver.0.1.0

* M5Stack ATOM 用直動機構を動かすためのクラスを追加

---

# Roadmap

* WiFiで操作するためのUIの追加 : (next ver.0.3.0)
* WiFiからペンプロッタのサーボの初期位置設定 : (next ver.0.4.0)
* WiFiからGコードなどの描画するデータを渡す機能 : (next ver.0.5.0)


src/main.cpp
/**
 * @file basic.ino
 * @author Akari (masiro.to.akari@gmail.com)
 * @brief
 * @version 0.1.0
 * @date 2023-05-01
 *
 * @copyright Copyright (c) 2023 / MaSiRo Project.
 *
 */
#if PLOT_PRIMARY || PLOT_REPLICA || PLOT_TEST

#include <ESP32Servo.h>
#include <M5Atom.h>

////////////////////////////////////////////////////////////
// PLOT_PRIMARY
////////////////////////////////////////////////////////////
#if PLOT_PRIMARY

#include "pen_plotter_primary.hpp"

PenPlotterPrimary pp;

void setup_pen_plotter()
{
    M5.dis.fillpix(CRGB::Yellow);
    pp.postion_start();
    M5.dis.fillpix(CRGB::Green);
}

void loop_pen_plotter()
{
    pp.loop();
}

////////////////////////////////////////////////////////////
// PLOT_REPLICA
////////////////////////////////////////////////////////////
#elif PLOT_REPLICA

#include "pen_plotter_replica.hpp"

PenPlotterReplica pp;

void setup_pen_plotter()
{
    M5.dis.fillpix(CRGB::Yellow);
    pp.begin();
    M5.dis.fillpix(CRGB::Black);
}

void loop_pen_plotter()
{
    pp.loop();
    M5.update();
    if (true == M5.Btn.wasReleased()) {
        pp.pen_release();
    }
}

////////////////////////////////////////////////////////////
// PLOT_PRIMARY[TEST]
////////////////////////////////////////////////////////////
#elif PLOT_TEST

#include "pen_plotter_primary.hpp"

PenPlotterPrimary pp;

void setup_pen_plotter()
{
    M5.dis.fillpix(CRGB::Yellow);
    if (0 < Serial.available()) {
        Serial.read();
    }
    pp.begin();
    M5.dis.fillpix(CRGB::Green);
}

void loop_pen_plotter()
{
    if (0 < Serial.available()) {
        log_d("==============================");
        unsigned int pos_min    = 0;
        unsigned int pos_max    = 100;
        unsigned int pos_middle = (unsigned int)(((double)abs((int)pos_max - (int)pos_min) / 2.0) + pos_min);

        uint16_t dat = Serial.read();
        log_d("  char: %c", dat);
        switch (dat) {
            case 'p':
                pp.pen_control(true);
                break;
            case 'o':
                pp.pen_control(false);
                break;
            case 'q':
                pp.move(pos_min, pos_min);
                break;
            case 'w':
                pp.move(pos_middle, pos_min);
                break;
            case 'e':
                pp.move(pos_max, pos_min);
                break;
            case 'a':
                pp.move(pos_min, pos_middle);
                break;
            case 's':
                pp.move(pos_middle, pos_middle);
                break;
            case 'd':
                pp.move(pos_max, pos_middle);
                break;
            case 'z':
                pp.move(pos_min, pos_max);
                break;
            case 'x':
                pp.move(pos_middle, pos_max);
                break;
            case 'c':
                pp.move(pos_max, pos_max);
                break;

            default:
                break;
        }
    }
}
#endif

////////////////////////////////////////////////////////////
// M5 ATOM
////////////////////////////////////////////////////////////

void setup()
{
    M5.begin(true, false, true);
    M5.dis.begin();
    ///////////////////////////////
    setup_pen_plotter();
    ///////////////////////////////
}

void loop()
{
    ///////////////////////////////
    loop_pen_plotter();
    ///////////////////////////////
}
#endif


src/pen_plotter_primary.cpp
/**
 * @file pen_plotter_primary.cpp
 * @author Akari (masiro.to.akari@gmail.com)
 * @brief
 * @version 0.2.0
 * @date 2023-05-03
 *
 * @copyright Copyright (c) 2023 / MaSiRo Project.
 *
 */
#include "pen_plotter_primary.hpp"

#include <Arduino.h>

PenPlotterPrimary::PenPlotterPrimary(int rx, int tx)
{
    this->_initialized = false;
    ///////////////////////////////
    // Linear Actuator
    ///////////////////////////////
    this->_lac.begin(1000);
    ///////////////////////////////
    // Serial
    ///////////////////////////////
    pinMode(rx, INPUT_PULLUP);
    pinMode(tx, INPUT_PULLUP);
    Serial2.begin(115200, SERIAL_8N1, rx, tx);
    Serial2.flush();
}

void PenPlotterPrimary::begin()
{
    do {
        Serial2.write('C');
        Serial2.write(this->_pen_position_up);
    } while (false == this->_move_replica_wait());
    this->_initialized = true;
    this->postion_start();
}
void PenPlotterPrimary::loop()
{
}

void PenPlotterPrimary::pen_control(bool flag_write)
{
    if (true == this->_initialized) {
        Serial2.write('P');
        if (true == flag_write) {
            Serial2.write(this->_pen_position_down);
        } else {
            Serial2.write(this->_pen_position_up);
        }
        this->_move_replica_wait();
    }
}

void PenPlotterPrimary::move(double x, double y)
{
    if (true == this->_initialized) {
        double a_x          = 0;
        double a_y          = 0;
        double current_x    = (double)this->pos_x;
        double current_y    = (double)this->pos_y;
        double difference_x = (x - this->pos_x);
        double difference_y = (y - this->pos_y);
        int max_value       = (int)(max(abs(difference_x), abs(difference_y)) / this->_resolution);

        if (0 != max_value) {
            a_x = (double)(difference_x / max_value);
            a_y = (double)(difference_y / max_value);
        }

        for (int i = 0; i < max_value; i++) {
            current_x += a_x;
            current_y += a_y;
            this->_move(current_x, current_y);
        }
        this->_move(x, y);
    }
}

void PenPlotterPrimary::postion_start()
{
    this->pen_control(false);
    this->_move(this->POSTION_START_X, this->POSTION_START_Y);
}

/////////////////////////////////////////////////
// private method
/////////////////////////////////////////////////
void PenPlotterPrimary::_move(double x, double y)
{
    bool result = this->_move_replica_start(x);
    if (true == result) {
        this->_move_replica_wait();
    }
    this->_move_primary(y);
}

void PenPlotterPrimary::_move_primary(double target)
{
    if (this->pos_y != target) {
        this->_lac.move(target);
        this->pos_y = target;
    }
}

bool PenPlotterPrimary::_move_replica_start(double target)
{
    bool result = false;
    if (this->pos_x != target) {
        unsigned int send_data = (unsigned int)(target * 100.0);
        result                 = true;
        Serial2.write('M');
        Serial2.write((send_data >> 8) & 0x00FF);
        Serial2.write(send_data & 0x00FF);
        this->pos_x = target;
    }
    return result;
}
bool PenPlotterPrimary::_move_replica_wait()
{
    bool result      = true;
    int data         = 0;
    int timeout_cont = (1000) * this->TIMEOUT_MS;
    do {
        if (0 < Serial2.available()) {
            data = Serial2.read();
        } else {
            delayMicroseconds(1);
        }
        if (0 >= timeout_cont) {
            result = false;
            break;
        }
        timeout_cont--;
    } while (data != 0xAA);
    return result;
}


src/pen_plotter_replica.cpp
/**
 * @file pen_plotter_replica.cpp
 * @author Akari (masiro.to.akari@gmail.com)
 * @brief
 * @version 0.2.0
 * @date 2023-05-03
 *
 * @copyright Copyright (c) 2023 / MaSiRo Project.
 *
 */
#include "pen_plotter_replica.hpp"

#include <Arduino.h>

PenPlotterReplica::PenPlotterReplica(int servo_wait_time_ms, int rx, int tx, int rc_servo_pin)
{
    this->_connected = false;
    ///////////////////////////////
    // RC Servo
    ///////////////////////////////
    this->_servo_wait = servo_wait_time_ms;
    pinMode(rc_servo_pin, OUTPUT);
    this->_pen.setPeriodHertz(50);
    this->_pen.attach(rc_servo_pin, 500, 2400);
    ESP32PWM::allocateTimer(0);
    ///////////////////////////////
    // Linear Actuator
    ///////////////////////////////
    this->_lac.begin(1000);
    ///////////////////////////////
    // Serial
    ///////////////////////////////
    pinMode(rx, INPUT_PULLUP);
    pinMode(tx, INPUT_PULLUP);
    Serial2.begin(115200, SERIAL_8N1, rx, tx);
    Serial2.flush();
}

/////////////////////////////////////////////////
// public method
/////////////////////////////////////////////////
void PenPlotterReplica::begin()
{
    // this->_pen.write(90);
    // this->_move_lac(50);
    do {
        this->loop();
    } while (false == this->_connected);
}

void PenPlotterReplica::loop()
{
    if (0 < Serial2.available()) {
        int timeout_count = TIMEOUT_MS * 1000;
        uint16_t pos;
        uint16_t dat = Serial2.read();

        if ('P' == dat) {
            while (0 >= Serial2.available()) {
                timeout_count--;
                if (0 >= timeout_count) {
                    break;
                }
                delayMicroseconds(1);
            }
            if (0 < timeout_count) {
                pos = Serial2.read();
                this->_move_pen(pos);
            }
        } else if ('M' == dat) {
            while (1 >= Serial2.available()) {
                timeout_count--;
                if (0 >= timeout_count) {
                    break;
                }
                delayMicroseconds(1);
            }
            if (0 < timeout_count) {
                dat = Serial2.read();
                pos = Serial2.read();

                double postion = (dat * 0x100 + pos) / 100.0;
                this->_move_lac(postion);
            }
        } else if ('C' == dat) {
            while (0 >= Serial2.available()) {
                timeout_count--;
                if (0 >= timeout_count) {
                    break;
                }
                delayMicroseconds(1);
            }
            if (0 < timeout_count) {
                pos = Serial2.read();
                this->_connection(pos);
            }
        } else {
            Serial2.read();
            timeout_count = 0;
        }
        if (0 >= timeout_count) {
            this->_fin();
            Serial2.clearWriteError();
        }
    }
}
void PenPlotterReplica::pen_release()
{
    this->_move_pen(this->_postion_pen_default);
}

/////////////////////////////////////////////////
// private method
/////////////////////////////////////////////////
void PenPlotterReplica::_move_pen(unsigned int postion)
{
    if (0 <= postion && postion <= 180) {
        this->_pen.write(postion);
        delay(this->_servo_wait);
    }
    this->_fin();
}
void PenPlotterReplica::_move_lac(double postion)
{
    this->_lac.move(postion);
    this->_fin();
}
void PenPlotterReplica::_connection(unsigned int postion)
{
    this->_postion_pen_default = postion;
    this->_connected           = true;
    this->_fin();
}

void PenPlotterReplica::_fin()
{
    Serial2.write(0xAA);
}


include/pen_plotter_primary.hpp
/**
 * @file pen_plotter_primary.hpp
 * @author Akari (masiro.to.akari@gmail.com)
 * @brief
 * @version 0.2.0
 * @date 2023-05-03
 *
 * @copyright Copyright (c) 2023 / MaSiRo Project.
 *
 */
#ifndef PEN_PLOTTER_PRIMARY_HPP
#define PEN_PLOTTER_PRIMARY_HPP

#include "linear_actuator_controller.hpp"

class PenPlotterPrimary {
public:
    PenPlotterPrimary(int rx = 32, int tx = 26);

public:
    void begin();
    void loop();

    void pen_control(bool flag_write);

    void postion_start();
    void move(double x, double y);

    double pos_x = 0;
    double pos_y = 0;

private:
    void _move(double x, double y);
    void _move_primary(double target);
    bool _move_replica_start(double target);
    bool _move_replica_wait();

private:
    LinearActuatorController _lac;

    int _pen_position_up   = 90;
    int _pen_position_down = 110;
    double _resolution     = 0.5;
    bool _initialized      = false;

private:
    const int TIMEOUT_MS      = 500;
    const int POSTION_START_X = 50;
    const int POSTION_START_Y = 25;
};

#endif


include/pen_plotter_replica.hpp
/**
 * @file pen_plotter_replica.hpp
 * @author Akari (masiro.to.akari@gmail.com)
 * @brief
 * @version 0.2.0
 * @date 2023-05-03
 *
 * @copyright Copyright (c) 2023 / MaSiRo Project.
 *
 */
#ifndef PEN_PLOTTER_REPLICA_HPP
#define PEN_PLOTTER_REPLICA_HPP

#include "linear_actuator_controller.hpp"

#include <ESP32Servo.h>

#ifndef RC_SERVO_PIN
#define RC_SERVO_PIN 25
#endif
#ifndef RC_SERVO_WAIT_TIME_MS
#define RC_SERVO_WAIT_TIME_MS 500
#endif

class PenPlotterReplica {
public:
    PenPlotterReplica(int servo_wait_time_ms = RC_SERVO_WAIT_TIME_MS, int rx = 26, int tx = 32, int rc_servo_pin = RC_SERVO_PIN);

public:
    void begin();
    void loop();
    void pen_release();

private:
    void _move_pen(unsigned int postion);
    void _move_lac(double postion);
    void _connection(unsigned int postion);
    void _fin();

private:
    LinearActuatorController _lac;

    int _servo_wait = RC_SERVO_WAIT_TIME_MS;
    Servo _pen;
    bool _connected                   = false;
    unsigned int _postion_pen_default = 90;

private:
    const int TIMEOUT_MS = 100;
};

#endif


include/linear_actuator_controller.hpp
/**
 * @file linear_actuator_controller.hpp
 * @author Akari (masiro.to.akari@gmail.com)
 * @brief
 * @version 0.1.0
 * @date 2023-05-01
 *
 * @copyright Copyright (c) 2023 / MaSiRo Project.
 *
 */
#ifndef LINEAR_ACTUATOR_CONTROLLER_HPP
#define LINEAR_ACTUATOR_CONTROLLER_HPP

#include <Arduino.h>

class LinearActuatorController {
public:
    /**
     * @brief Construct a new Linear Actuator Controller object
     *
     * @param mt_driver_1
     * @param mt_driver_2
     * @param adc
     */
    LinearActuatorController(uint8_t mt_driver_1 = 22, //
                             uint8_t mt_driver_2 = 19,
                             uint8_t adc         = 33);

    /**
     * @brief Set the up margin object
     *
     * @param inside_percentage
     * @param outside_percentage
     */
    void setup_margin(double inside_percentage, double outside_percentage);
    /**
     * @brief
     *
     * @param timeout_ms time out unit:[ms]
     */
    void begin(uint16_t timeout_ms = (60 * 1000));
    /**
     * @brief move linear actuator
     *
     * @param percentage
     */
    bool move(double percentage);
    /**
     * @brief Get the position
     *      unit:[%]
     *
     * @return double value
     */
    double get_position_percentage();

private:
    typedef enum
    {
        MODE_LINEAR_ACTUATOR_NOT_MOVING,
        MODE_LINEAR_ACTUATOR_INSIDE,
        MODE_LINEAR_ACTUATOR_OUTSIDE,
    } MODE_LINEAR_ACTUATOR_T;
    uint16_t _read_position(uint8_t pin_no);

private:
    uint8_t _mt_driver_1 = 0;
    uint8_t _mt_driver_2 = 0;
    uint8_t _adc         = 0;
    uint16_t _pos        = 0;

    int8_t _samples_analog = 16;
    uint16_t _margin_min;
    uint16_t _margin_max;

    uint32_t _timeout_count_max = 0xFFFFFFFF;

    bool _initialized = false;

private:
    const uint8_t MOTOR_STOP = LOW;
    const uint8_t MOTOR_RUN  = HIGH;
};

#endif


src/linear_actuator_controller.cpp
/**
 * @file linear_actuator_controller.cpp
 * @author Akari (masiro.to.akari@gmail.com)
 * @brief
 * @version 0.1.0
 * @date 2023-05-01
 *
 * @copyright Copyright (c) 2023 / MaSiRo Project.
 *
 */
#include "linear_actuator_controller.hpp"

#define POS_MIN (0)
#ifdef ANALOG_10_BIT
#define POS_MAX (1023)
#else
#define POS_MAX (4095)
#endif

LinearActuatorController::LinearActuatorController(uint8_t mt_driver_1, uint8_t mt_driver_2, uint8_t adc)
{
    this->_initialized = false;
    this->_mt_driver_1 = mt_driver_1;
    this->_mt_driver_2 = mt_driver_2;
    this->_adc         = adc;
    this->setup_margin(5, 95);
    pinMode(this->_adc, INPUT); // ADC
}

void LinearActuatorController::setup_margin(double inside_percentage, double outside_percentage)
{
    this->_margin_min = (uint16_t)((POS_MAX - POS_MIN) * (max(0.0, min(100.0, inside_percentage))) / 100.0);
    this->_margin_max = (uint16_t)((POS_MAX - POS_MIN) * (max(0.0, min(100.0, outside_percentage))) / 100.0);
}

void LinearActuatorController::begin(uint16_t timeout_ms)
{
    pinMode(this->_mt_driver_1, OUTPUT);                // Motor Driver[IN1]
    pinMode(this->_mt_driver_2, OUTPUT);                // Motor Driver[IN2]
    digitalWrite(this->_mt_driver_1, this->MOTOR_STOP); // Motor Stop[IN1]
    digitalWrite(this->_mt_driver_2, this->MOTOR_STOP); // Motor Stop[IN2]

    this->_timeout_count_max = (timeout_ms * 1000) / this->_samples_analog;
    this->_initialized       = true;
    this->_pos               = this->_read_position(this->_adc);
}

bool LinearActuatorController::move(double percentage)
{
    bool result = true;
    if (true == this->_initialized) {
        MODE_LINEAR_ACTUATOR_T mode;
        uint32_t timeout = this->_timeout_count_max;
        uint16_t target  = 0;
        if (percentage < 0) {
            percentage = 0;
        } else if (percentage > 100) {
            percentage = 100;
        }
        target = (uint16_t)((POS_MAX - POS_MIN) * (percentage / 100.0));

        if (target > this->_pos) {
            mode = MODE_LINEAR_ACTUATOR_OUTSIDE;
            digitalWrite(this->_mt_driver_1, this->MOTOR_RUN);
            digitalWrite(this->_mt_driver_2, this->MOTOR_STOP);
            log_d(" Postion : OUTSIDE : %d -> %d[%0.3f%%]", this->_pos, target, percentage);
        } else if (target < this->_pos) {
            mode = MODE_LINEAR_ACTUATOR_INSIDE;
            digitalWrite(this->_mt_driver_1, this->MOTOR_STOP);
            digitalWrite(this->_mt_driver_2, this->MOTOR_RUN);
            log_d(" Postion : INSIDE  : %d -> %d[%0.3f%%]", this->_pos, target, percentage);
        } else {
            mode = MODE_LINEAR_ACTUATOR_NOT_MOVING;
            log_d(" NOT MOVING");
        }

        while (MODE_LINEAR_ACTUATOR_NOT_MOVING != mode) {
            this->_pos = this->_read_position(this->_adc);
            log_v(" MOVE : %s[%d -> %d]", (MODE_LINEAR_ACTUATOR_INSIDE == mode) ? "INSIDE " : "OUTSIDE", this->_pos, target);

            if ((MODE_LINEAR_ACTUATOR_OUTSIDE == mode) && (target <= this->_pos)) {
                mode = MODE_LINEAR_ACTUATOR_NOT_MOVING;
                break;
            } else if ((MODE_LINEAR_ACTUATOR_INSIDE == mode) && (target >= this->_pos)) {
                mode = MODE_LINEAR_ACTUATOR_NOT_MOVING;
                break;
            }
            if (0 > timeout) {
                result = false;
                break;
            }
            timeout--;
        }
        // Motor Brake
        digitalWrite(this->_mt_driver_1, this->MOTOR_STOP);
        digitalWrite(this->_mt_driver_2, this->MOTOR_STOP);
    } else {
        result = false;
    }
    return result;
}

double LinearActuatorController::get_position_percentage()
{
    uint16_t pos      = this->_read_position(this->_adc);
    double percentage = ((double)(pos - POS_MIN) / (double)(POS_MAX - POS_MIN)) * 100.0;
    return percentage;
}

////////////////////////////////////////////////////////////////////////////
// Private Methods
////////////////////////////////////////////////////////////////////////////

uint16_t LinearActuatorController::_read_position(uint8_t pin_no)
{
    uint16_t pos     = 0;
    uint16_t max_pos = POS_MIN;
    uint16_t min_pos = POS_MAX;

    for (int8_t i = 0; i < this->_samples_analog; i++) {
        uint16_t dat = analogRead(pin_no);
        pos += dat;
        max_pos = max(max_pos, dat);
        min_pos = min(min_pos, dat);
        delayMicroseconds(1);
    }
    //////////////////////
    // Exclude outliers
    pos -= max_pos;
    pos -= min_pos;
    //////////////////////
    pos = pos / (this->_samples_analog - 2);
    if (pos < this->_margin_min) {
        pos = this->_margin_min;
    } else if (pos > this->_margin_max) {
        pos = this->_margin_max;
    }
    pos = map(pos, this->_margin_min, this->_margin_max, POS_MIN, POS_MAX);

    return pos;
}
